core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Workflow.scala

```diff
@@ -150,177 +55,166 @@ class Workflow(
 
   def getPythonWorkers: Iterable[ActorVirtualIdentity] =
-    workerToOperatorExec
-      .filter({
-        case (_: ActorVirtualIdentity, operatorExecutor: IOperatorExecutor) =>
-          operatorExecutor.isInstanceOf[PythonUDFOpExecV2]
-      }) map { case (workerId: ActorVirtualIdentity, _: IOperatorExecutor) => workerId }
+    workerToOpExecConfig.filter(worker => worker._2.opExecClass == classOf[PythonUDFOpExecV2]).keys
 
-  def getOperatorToWorkers: Iterable[(OperatorIdentity, Seq[ActorVirtualIdentity])] = {
-    getAllOperatorIds.map(opId => {
+  def getOperatorToWorkers: Iterable[(LayerIdentity, Seq[ActorVirtualIdentity])] = {
+    physicalPlan.allOperatorIds.map(opId => {
       (opId, getAllWorkersForOperators(Array(opId)).toSeq)
     })
   }
+
   def getAllWorkersForOperators(
-      operators: Array[OperatorIdentity]
+      operators: Array[LayerIdentity]
   ): Array[ActorVirtualIdentity] = {
-    operators.flatMap(opId => operatorToOpExecConfig(opId).getAllWorkers)
+    operators.flatMap(opId => physicalPlan.operatorMap(opId).getAllWorkers)
   }
 
-  def getPythonOperators(fromOperatorsList: Array[OperatorIdentity]): Array[OperatorIdentity] = {
+  def getPythonOperators(fromOperatorsList: Array[LayerIdentity]): Array[LayerIdentity] = {
     fromOperatorsList.filter(opId =>
-      operatorToOpExecConfig(opId).getAllWorkers.size > 0 && operatorToOpExecConfig(
-        opId
-      ).getAllWorkers.forall(wid => workerToOperatorExec(wid).isInstanceOf[PythonUDFOpExecV2])
+      physicalPlan.operatorMap(opId).getAllWorkers.nonEmpty &&
+        physicalPlan.operatorMap(opId).opExecClass == classOf[PythonUDFOpExecV2]
     )
   }
 
   def getPythonWorkerToOperatorExec(
-      pythonOperators: Array[OperatorIdentity]
+      pythonOperators: Array[LayerIdentity]
   ): Iterable[(ActorVirtualIdentity, PythonUDFOpExecV2)] = {
-    val allWorkers = pythonOperators
-      .map(opId => operatorToOpExecConfig(opId).getAllWorkers)
-      .flatten
-    workerToOperatorExec
-      .filter({
-        case (id: ActorVirtualIdentity, operatorExecutor: IOperatorExecutor) =>
-          allWorkers.contains(id)
-      })
-      .asInstanceOf[Iterable[(ActorVirtualIdentity, PythonUDFOpExecV2)]]
+    throw new NotImplementedError("fix python")
+//    val allWorkers = pythonOperators.flatMap(opId => physicalPlan.operatorMap(opId).getAllWorkers)
+//    workerToOperatorExec
+//      .filter({
+//        case (id: ActorVirtualIdentity, operatorExecutor: IOperatorExecutor) =>
+//          allWorkers.contains(id)
+//      })
+//      .asInstanceOf[Iterable[(ActorVirtualIdentity, PythonUDFOpExecV2)]]
   }
 
   def getPythonWorkerToOperatorExec: Iterable[(ActorVirtualIdentity, PythonUDFOpExecV2)] =
-    workerToOperatorExec
+    workerToOpExecConfig
       .filter({
         case (_: ActorVirtualIdentity, operatorExecutor: IOperatorExecutor) =>
           operatorExecutor.isInstanceOf[PythonUDFOpExecV2]
```

Comment:
Can we cast `NewOpExecConfig` to `IOperatorExecutor`?

Reply:
no, this has been fixed